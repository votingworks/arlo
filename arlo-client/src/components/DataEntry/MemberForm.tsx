import React from 'react'
import { H1, RadioGroup, Radio } from '@blueprintjs/core'
import { Formik, FormikProps, Form, getIn } from 'formik'
import FormSection from '../Atoms/Form/FormSection'
import { IAuditBoardMember } from '../../types'
import FormButton from '../Atoms/Form/FormButton'
import { LabelText, NameField } from './Atoms'

interface IProps {
  boardName: string
  jurisdictionName: string
  submitMembers: (members: IAuditBoardMember[]) => void
}

const MemberForm: React.FC<IProps> = ({
  boardName,
  jurisdictionName,
  submitMembers,
}: IProps) => {
  return (
    <div>
      <H1>{boardName}: Member Sign-in</H1>
      <p>
        Enter the information below for each member of {jurisdictionName}{' '}
        {boardName} below, then click &quot;Next&quot; to proceed.
      </p>
      <Formik
        initialValues={[
          {
            name: '',
            affiliation: '',
          },
          {
            name: '',
            affiliation: '',
          },
        ]}
        onSubmit={members =>
          submitMembers(
            members
              .filter(({ name }) => name)
              .map(({ name, affiliation }) => ({
                name,
                affiliation: affiliation || null,
              }))
          )
        }
        render={({
          setFieldValue,
          values,
          handleSubmit,
        }: FormikProps<[IAuditBoardMember, IAuditBoardMember]>) => (
          <Form>
            {[0, 1].map(i => (
              <FormSection label="Audit Board Member" key={i}>
                {/* eslint-disable jsx-a11y/label-has-associated-control */}
                <LabelText htmlFor={`[${i}]name`}>Full Name</LabelText>
                <NameField name={`[${i}]name`} id={`[${i}]name`} />
                <LabelText htmlFor={`[${i}]affiliation`}>
                  Party Affiliation <i>(optional)</i>
                </LabelText>
                <RadioGroup
                  name={`[${i}]affiliation`}
                  onChange={e =>
                    setFieldValue(`[${i}]affiliation`, e.currentTarget.value)
                  }
                  selectedValue={getIn(values, `[${i}]affiliation`)}
                >
                  <Radio value="DEM">Democrat</Radio>
                  <Radio value="REP">Republican</Radio>
                  <Radio value="LIB">Libertarian</Radio>
                  <Radio value="IND">Independent/Unaffiliated</Radio>
                  <Radio value="OTH">Other</Radio>
                  <Radio value="">None</Radio>
                </RadioGroup>
              </FormSection>
            ))}
            <FormButton
              intent="primary"
              type="button"
              disabled={!(values[0].name || (values[0].name && values[1].name))}
              onClick={handleSubmit}
            >
              Next
            </FormButton>
          </Form>
        )}
      />
    </div>
  )
}

export default MemberForm
