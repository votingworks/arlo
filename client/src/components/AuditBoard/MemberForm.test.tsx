import React from 'react'
import { render, fireEvent, waitFor } from '@testing-library/react'
import MemberForm from './MemberForm'

describe('MemberForm', () => {
  it('renders correctly', () => {
    const { container } = render(
      <MemberForm
        boardName="board name"
        jurisdictionName="jurisdiction name"
        submitMembers={jest.fn()}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('handles inputs', async () => {
    const submitMembersMock = jest.fn()
    const { queryAllByLabelText, queryAllByText, getByText } = render(
      <MemberForm
        boardName="board name"
        jurisdictionName="jurisdiction name"
        submitMembers={submitMembersMock}
      />
    )

    const names = queryAllByLabelText('Full Name') as HTMLInputElement[]
    const parties = queryAllByText('Democrat') as HTMLInputElement[]
    expect(names.length).toBe(2)
    expect(parties.length).toBe(2)
    names.forEach(nameInput => {
      fireEvent.change(nameInput, { target: { value: 'my name' } })
      expect(nameInput.value)
    })
    parties.forEach(partyInput => {
      fireEvent.click(partyInput, { bubbles: true })
    })

    const nextButton = getByText('Next')
    fireEvent.click(nextButton, { bubbles: true })
    await waitFor(() => {
      expect(submitMembersMock).toBeCalledTimes(1)
    })
  })
})
