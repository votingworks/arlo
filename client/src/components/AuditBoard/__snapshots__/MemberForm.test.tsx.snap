// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberForm renders correctly 1`] = `
<div>
  <div>
    <h1
      class="bp3-heading"
    >
      board name
      : Member Sign-in
    </h1>
    <p>
      Enter the information below for each member of 
      jurisdiction name
       
      board name
       below, then click "Next" to proceed.
    </p>
    <form>
      <div
        class="sc-bdVaJa kKNzTG"
      >
        <h5
          class="bp3-heading"
        >
          Audit Board Member
        </h5>
        <label
          class="sc-dnqmqq kSLuZR"
          for="[0]name"
        >
          Full Name
        </label>
        <input
          class="sc-iwsKbI hdoOFn"
          id="[0]name"
          name="[0]name"
          value=""
        />
        <label
          class="sc-dnqmqq kSLuZR"
          for="[0]affiliation"
        >
          Party Affiliation 
          <i>
            (optional)
          </i>
        </label>
        <div>
          <label
            class="bp3-control bp3-radio"
          >
            <input
              name="[0]affiliation"
              type="radio"
              value="DEM"
            />
            <span
              class="bp3-control-indicator"
            />
            Democrat
          </label>
          <label
            class="bp3-control bp3-radio"
          >
            <input
              name="[0]affiliation"
              type="radio"
              value="REP"
            />
            <span
              class="bp3-control-indicator"
            />
            Republican
          </label>
          <label
            class="bp3-control bp3-radio"
          >
            <input
              name="[0]affiliation"
              type="radio"
              value="LIB"
            />
            <span
              class="bp3-control-indicator"
            />
            Libertarian
          </label>
          <label
            class="bp3-control bp3-radio"
          >
            <input
              name="[0]affiliation"
              type="radio"
              value="IND"
            />
            <span
              class="bp3-control-indicator"
            />
            Independent/Unaffiliated
          </label>
          <label
            class="bp3-control bp3-radio"
          >
            <input
              name="[0]affiliation"
              type="radio"
              value="OTH"
            />
            <span
              class="bp3-control-indicator"
            />
            Other
          </label>
          <label
            class="bp3-control bp3-radio"
          >
            <input
              checked=""
              name="[0]affiliation"
              type="radio"
              value=""
            />
            <span
              class="bp3-control-indicator"
            />
            None
          </label>
        </div>
      </div>
      <div
        class="sc-bdVaJa kKNzTG"
      >
        <h5
          class="bp3-heading"
        >
          Audit Board Member
        </h5>
        <label
          class="sc-dnqmqq kSLuZR"
          for="[1]name"
        >
          Full Name
        </label>
        <input
          class="sc-iwsKbI hdoOFn"
          id="[1]name"
          name="[1]name"
          value=""
        />
        <label
          class="sc-dnqmqq kSLuZR"
          for="[1]affiliation"
        >
          Party Affiliation 
          <i>
            (optional)
          </i>
        </label>
        <div>
          <label
            class="bp3-control bp3-radio"
          >
            <input
              name="[1]affiliation"
              type="radio"
              value="DEM"
            />
            <span
              class="bp3-control-indicator"
            />
            Democrat
          </label>
          <label
            class="bp3-control bp3-radio"
          >
            <input
              name="[1]affiliation"
              type="radio"
              value="REP"
            />
            <span
              class="bp3-control-indicator"
            />
            Republican
          </label>
          <label
            class="bp3-control bp3-radio"
          >
            <input
              name="[1]affiliation"
              type="radio"
              value="LIB"
            />
            <span
              class="bp3-control-indicator"
            />
            Libertarian
          </label>
          <label
            class="bp3-control bp3-radio"
          >
            <input
              name="[1]affiliation"
              type="radio"
              value="IND"
            />
            <span
              class="bp3-control-indicator"
            />
            Independent/Unaffiliated
          </label>
          <label
            class="bp3-control bp3-radio"
          >
            <input
              name="[1]affiliation"
              type="radio"
              value="OTH"
            />
            <span
              class="bp3-control-indicator"
            />
            Other
          </label>
          <label
            class="bp3-control bp3-radio"
          >
            <input
              checked=""
              name="[1]affiliation"
              type="radio"
              value=""
            />
            <span
              class="bp3-control-indicator"
            />
            None
          </label>
        </div>
      </div>
      <button
        class="bp3-button bp3-disabled bp3-intent-primary sc-bxivhb WkpPV"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="bp3-button-text"
        >
          Next
        </span>
      </button>
    </form>
  </div>
</div>
`;
